### PR TITLE
BRMS-425: authentication changes

### PR DESCRIPTION
**Pull request by chris-ianson**
@adamconder

### CODE REVIEW

File: app/uk/gov/hmrc/brm/config/microserviceGlobal.scala [hunk 1]
```diff
@@ -34,26 +34,31 @@ object ControllerConfiguration extends ControllerConfig {
 
 object GROConnectorConfiguration extends ServicesConfig {
 
+  private lazy val tlsConfigPath = "birth-registration-matching.gro.v1.tls."
```
Comment (adamconder) [inline app/uk/gov/hmrc/brm/config/microserviceGlobal.scala]: Doesn't need to be lazy.

File: app/uk/gov/hmrc/brm/config/microserviceGlobal.scala [hunk 2]
```diff
@@ -34,26 +34,31 @@ object ControllerConfiguration extends ControllerConfig {
...
-  lazy val tlsPrivateKeystoreKey = getConfString("birth-registration-matching.privateKeystoreKey", throw new RuntimeException(message("privateKeystoreKey")))
-  lazy val hostname = getConfBool("birth-registration-matching.allowHostnameMismatch", throw new RuntimeException(message("allowHostnameMismatch")))
-  lazy val tlsVersion = getConfString("birth-registration-matching.tlsVersion", throw new RuntimeException(message("tlsVersion")))
-  lazy val tlsEnabled = getConfBool("birth-registration-matching.tlsEnabled", throw new RuntimeException(message("tlsEnabled")))
-  lazy val connectionTimeout = getConfInt("birth-registration-matching.connectionTimeout",throw new RuntimeException(message("connectionTimeout")))
-  lazy val readTimeout = getConfInt("birth-registration-matching.readTimeout",throw new RuntimeException(message("readTimeout")))
-  lazy val delayAttemptInMilliseconds : DelayTime = getConfInt("birth-registration-matching.delayAttemptInMilliseconds",
+
+  lazy val username = getConfString(authenticationConfigPath + "username",throw new RuntimeException(message("username")))
```
Comment (adamconder) [inline app/uk/gov/hmrc/brm/config/microserviceGlobal.scala]: Can you do string interpolation instead. s"$authenticationConfigPath$username"

File: conf/application.conf [hunk 3]
```diff
@@ -146,31 +146,52 @@ microservice {
             host = localhost
             port = 8500
         }
+
         birth-registration-matching {
+
+            gro  {
+                v1 {
+                    tls {
```
Comment (adamconder) [inline conf/application.conf]: @chris-ianson Move this out of v1 config, can this be it's own object.

File: conf/application.conf [hunk 4]
```diff
@@ -146,31 +146,52 @@ microservice {
...
+                        tlsEnabled = false
+                        readTimeout = 19500
+                        connectionTimeout = 5500
+                        privateKeystore = privateKeystore
+                        privateKeystoreKey = privateKeystoreKey
+                        certificateExpiryDate = 2018-06-28
+                    }
+
+                    authentication {
+                        delayAttemptInMilliseconds = 5000
```
Comment (adamconder) [inline conf/application.conf]: Move the attempts and the delay attemps into the gro object.

File: conf/application.conf [hunk 5]
```diff
@@ -146,31 +146,52 @@ microservice {
...
+
+                v2 {
+                    authentication {
+                        delayAttemptInMilliseconds = 5000
+                        delayAttempts = 3
+                        key = key
+                        username = username
+                        clientID = clientID
+                        clientSecret = clientSecret
+                        grantType = grantType
```
Comment (adamconder) [inline conf/application.conf]: Set this to password

File: conf/application.conf [hunk 6]
```diff
@@ -150,26 +150,28 @@ microservice {
...
+                v1 {
                     authentication {
-                        delayAttemptInMilliseconds = 5000
-                        delayAttempts = 3
                         key = key
                         username = username
                     }
                 }
 
+                delayAttemptInMilliseconds = 5000
```
Comment (adamconder) [inline conf/application.conf]: Can you wrap these in a http object and also move readTimeout and connectionTimeout into this.

File: conf/application.conf [hunk 7]
```diff
@@ -150,26 +150,28 @@ microservice {
...
+                    allowHostnameMismatch = false
+                    tlsEnabled = false
+                    readTimeout = 19500
+                    connectionTimeout = 5500
+                    privateKeystore = privateKeystore
+                    privateKeystoreKey = privateKeystoreKey
+                    certificateExpiryDate = 2018-06-28
+                }
+
+                v1 {
```
Comment (adamconder) [inline conf/application.conf]: Make this authentication / v1 || v2

**Comment by chris-ianson** [comment]
@adamconder I've added in client config and grant type to support latest GRO authentication. The old authentication has been replaced due to Daniel's comments about removing this altogether. 

Config keys have been updated and will have to possibly be changed

* Jenkins config
* Acceptance test readme
* Any internal documentation that specifies the old password

The stub has one small change to content-type validation and will need to be merged before merging this. Once we're happy with this we can update each environments config, deploy and test.

File: conf/application.conf [hunk 8]
```diff
@@ -146,31 +146,49 @@ microservice {
...
+                    allowHostnameMismatch = false
+                    tlsEnabled = false
+                    privateKeystore = privateKeystore
+                    privateKeystoreKey = privateKeystoreKey
+                    certificateExpiryDate = 2018-06-28
+                }
+
+                authentication {
+
+                    v2 {
```
Comment (chris-ianson) [inline conf/application.conf]: A bit picky, but does this really need to be wrapped in a v2 block? I'm not sure now.

File: test/uk/gov/hmrc/brm/connectors/BirthConnectorSpec.scala [hunk 9]
```diff
@@ -126,9 +126,9 @@ class BirthConnectorSpec extends UnitSpec with MockitoSugar with BeforeAndAfter
 
       implicit val metrics = mock[BRMMetrics]
 
-      "BirthErrorResponse of 4xx when authentication returns BadRequest" in {
+      "return 4xx when authentication returns BadRequest" in {
         val authResponse = Response.apply(
```
Comment (adamconder) [inline test/uk/gov/hmrc/brm/connectors/BirthConnectorSpec.scala]: @chris-ianson This doesn't assert anything?

**Comment by chris-ianson** [comment]
@adamconder Added new host for authentication.

Think this is ready to merge but we also have to review app-config-base and service-manager-config

**Pull request merged**